### ISSUE TITLE
Feature request: Allow plugins to define their own symbolic names for use with extends

### ISSUE THREAD

**Issue by jimbolla** [eslint/eslint#3659]
As requested in yannickcr/eslint-plugin-react#192 it would be useful for plugins to define their own symbolic names for configs. Example:

    {
        "extends": [ "eslint:recommended", "eslint-plugin-react:recommended" ],
    }

Currently, "eslint:recommended" is a [hardcoded value in config.js](https://github.com/eslint/eslint/blob/02c6c848d7d70c285269c54718dbc9c12d3f0720/lib/config.js#L151). and there's no way to add additional config names.The way I envision expanding this functionality would be to make this into a map that plugins would be able to add to when they are loaded.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Related: #3458 

That's an interesting idea! How would you envision defining those configs in the plugin?

**Comment by jimbolla** [comment]
I assume it'd be another property in the plugin definition:

    var strict = require(

    module.exports = {
        rules: { /* ... */ },
        extendsAliases: {
            "strict": require("./conf/strict.json"),
            "recommended": require("./conf/recommended.json")
        }
    };

The convention could be that callsite prepends the plugin name and colon so that using it would look like "eslint-plugin-something:strict" or "eslint-plugin-something:recommended".

**Comment by nzakas** [comment]
Ah I see. I think I'd just call that key `configs`, but otherwise, I like it.

@eslint/eslint-team thoughts?

**Comment by ilyavolodin** [comment]
Just to clarify, we are talking about `eslint-plugin-` not `eslint-config-` right? I don't have a problem adding this in, I'm just wondering why we should keep `eslint-config-` around if this going to be added to plugins. Seems like we would have two separate ways to get the same exact result.

**Comment by nzakas** [comment]
Yeah, we would probably deprecate shareable configs as a standalone thing, but we can't remove support for them.

**Comment by nzakas** [comment]
@jimbolla are you interested in submitting a pull request for this?

**Comment by jimbolla** [comment]
Sorry, I doubt I could implement this any timely manner.

**Comment by jimbolla** [comment]
I actually started working on this, but when I realized EVERY method requires jsdoc annotations, I was immediately turned off. I think require-jsdoc has the unintended consequence of discouraging good function decomposition.

**Comment by nzakas** [comment]
In a project as large as this with as many contributors as we have, we just can't afford to have undocumented functions. I don't see any way in which requiring documentation comments for functions discourages any best practices.

**Comment by platinumazure** [comment]
Is anyone on this?

I have a few things pending on my end (still need to get the `indent` PR done, for one thing), but if nobody picks this up in a week or so, I'd be interested.

That said, I'm not really sure where to start-- the proposed API isn't really clear. I can't tell if plugins are supposed to export configuration options or merely reference them and have some sort of magic go on behind the scenes.

**Comment by ilyavolodin** [comment]
I was planning on looking into this once @nzakas merges changes to plugins that are currently pending.

**Comment by platinumazure** [comment]
Sounds good @ilyavolodin, I'll leave it to you then.

**Comment by nzakas** [comment]
Just merged.

**Comment by ilyavolodin** [comment]
I started working on this, but pretty much instantly ran into a problem. Right now we allow users to omit `eslint-config-` when they use extend. With addition of support for configs in plugins, it's not clear how we can resolve omited prefix.
One thing I can think of is that we will only allow named configs in plugins (i.e. `eslint-plugin-react:strict`) then column will act as identifier for plugin vs. config. Would that work?

**Comment by ilyavolodin** [comment]
I'm also not sure how we should handle scoped packages. I assume we want to support them as well, but the way we parse them now is pretty complex and I'm not sure I can make it work for both plugins and configs at the same time.

**Comment by btmills** [comment]
So before, if I had extends `btmills/es6`, that would map to `eslint-config-btmills/es6` to specify the es6 variant of my custom config. Now, I convert my config to a plugin. To resolve the difference, I instead extend `btmills:es6`, which maps to a specific `configs.es6` export of `eslint-plugin-btmills`. Other possibility: `@btmills:es6` becomes `require("@btmills/eslint-plugin").configs.es6`? I think that works, and I even like the syntax. I think the only think this couldn't do is have plugins have a default exported config because there'd be now way to disambiguate between a config module and a plugin module. Would it be too much to try eslint-config-btmills first and if that doesn't exist then eslint-plugin-btmills.configs.default?

**Comment by ilyavolodin** [comment]
I guess we could do the whole "try to load config first and if it fails load plugin instead" but I feel that might be very slow. I think we load configs a lot during the course of linting, and hitting this `try/catch` many times would be costly (although would make it very easy to implement this feature).

**Comment by nzakas** [comment]
We could try to make this more explicit. So for shareable configs:

```
extends: "btmills/es6"
```

For plugin configs:

```
extends: "plugin:btmills/es6`
```

That way, this behaves similar to how we reference environments and rules in plugins, and we can treat scoped packages in a similar manner:

```
extends: "plugin:@btmills/myplugin/es6`
```

?

**Comment by ilyavolodin** [comment]
@nzakas You mentioned before that once this is implemented, we would deprecate shareable configs. It would be strange if implicit way would be deprecated...

**Comment by nzakas** [comment]
@ilyavolodin Yeah, I think I was wrong about that. There's a big ecosystem of shareable configs, so I don't think we can ever really deprecate them. We can, however, add new ways to load configs.

**Comment by ilyavolodin** [comment]
Ok, sounds good. So in that case, I will be branching code based on `plugin:` prefix

**Comment by ilyavolodin** [comment]
More questions. Do we want to support default config in plugins? Something like:
```js
extends: "plugin@btmills/myplugin
```
We can assume that config inside a plugin will always use rules from that plugin, right? Should we still require that the current plugin be added under `plugins` section of the config?

**Comment by nzakas** [comment]
Let's hold off on default configs for now. 

> We can assume that config inside a plugin will always use rules from that plugin, right? 

Not necessarily. This isn't required. It could also be a plugin without rules. 

>  Should we still require that the current plugin be added under plugins section of the config?

I'm assuming you mean "should a config in a plugin have to mention that plugin in the plugins array?" The answer is no, because you cannot use a config from a plugin without that plugin appearing in the plugins array of the config using extends.
